doc/hr.1
.\" Automatically generated by Pandoc 3.0.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "HR" "1" "" "hr-haskell 0.7.1.0 (2026-01-10)" "hr Manual"
.nh
.SH NAME
.PP
\f[V]hr\f[R] - horizontal rule for the terminal
.SH SYNOPSIS
.PP
\f[V]hr\f[R] [\f[I]OPTIONS\f[R]] [NOTE \&...]
.SH DESCRIPTION
.PP
\f[V]hr\f[R] is a utility for displaying a horizontal rule in a
terminal.
.SH OPTIONS
.TP
-h, --help
show help and exit
.TP
--version
show version and exit
.TP
-w, --width \f[I]CHARS\f[R]
target rule width (default: terminal width)
.TP
-d, --default \f[I]CHARS\f[R]
default target rule width (default: 80)
.TP
-a, --ascii
use ASCII lines (default: use Unicode lines)
.TP
-t, --time
show time
.TP
-f, --format \f[I]FORMAT\f[R]
time format (default: \f[V]%Y-%m-%d %H:%M:%S\f[R])
.RS
.PP
The following format codes are supported:
.IP \[bu] 2
\f[V]%Y\f[R] - four-digit year
.IP \[bu] 2
\f[V]%y\f[R] - two-digit year
.IP \[bu] 2
\f[V]%m\f[R] - two-digit month
.IP \[bu] 2
\f[V]%d\f[R] - two-digit day
.IP \[bu] 2
\f[V]%H\f[R] - two-digit hour using 24-hour clock
.IP \[bu] 2
\f[V]%I\f[R] - two-digit hour using 12-hour clock
.IP \[bu] 2
\f[V]%p\f[R] - locale equivalent of AM or PM
.IP \[bu] 2
\f[V]%M\f[R] - two-digit minute
.IP \[bu] 2
\f[V]%S\f[R] - two-digit second
.IP \[bu] 2
\f[V]%q\f[R] - twelve-digit picosecond
.IP \[bu] 2
\f[V]%z\f[R] - UTC offset
.RE
.TP
-i, --input
read note from STDIN within MS milliseconds
.TP
--timeout \f[I]MS\f[R]
timeout in milliseconds (default: 500)
.SH ARGUMENTS
.TP
\f[I]NOTE\f[R]
show a note
.SH EXIT CODES
.TP
0
no error
.TP
2
command-line error
.SH PROJECT
.TP
GitHub:
<https://github.com/ExtremaIS/hr-haskell>
.TP
Reporting issues:
GitHub: <https://github.com/ExtremaIS/hr-haskell/issues>
.RS
.PP
Email: <bugs@extrema.is>
.RE
.TP
Copyright
Copyright (c) 2019-2026 Travis Cardwell
.TP
License
The MIT License <https://opensource.org/licenses/MIT>


CHANGELOG.md
# `hr-haskell` Changelog

This project follows the [Haskell package versioning policy][PVP], with
versions in `A.B.C.D` format.  `A` may be incremented arbitrarily for
non-technical reasons, but [semantic versioning][SemVer] is otherwise
followed, where `A.B` is the major version, `C` is the minor version, and `D`
is the patch version.  Initial development uses versions `0.0.C.D`, for which
every version is considered breaking.

[PVP]: <https://pvp.haskell.org/>
[SemVer]: <https://semver.org/>

The format of this changelog is based on [Keep a Changelog][KaC], with the
following conventions:

* Level-two heading `Unreleased` is used to track changes that have not been
  released.
* Other level-two headings specify the release in `A.B.C.D (YYYY-MM-DD)`
  format, with newer versions above older versions.
* Level-three headings are used to categorize changes as follows:
    1. Breaking
    2. Non-Breaking
* Changes are listed in arbitrary order and present tense.

[KaC]: <https://keepachangelog.com/en/1.0.0/>

## Unreleased

## 0.7.1.0 (2026-01-10)

### Non-Breaking

* Bump `base` dependency version upper bound
* Bump `optparse-applicative` dependency version upper bound
* Bump `time` dependency version upper bound
* Remove mock tests, vendored `HMock` and `explainable-predicates`
  dependencies

## 0.7.0.0 (2024-12-04)

### Breaking

* Remove support for GHC 8.6, constraining lower bounds
* Remove support for GHC 8.4, constraining lower bounds
* Remove support for GHC 8.2, constraining lower bounds
* Change minimal Cabal from 1.24 to 3.0

### Non-Breaking

* Bump `base` dependency version upper bound
* Bump `tasty` dependency version upper bound
* Bump `text` dependency version upper bound
* Bump `time` dependency version upper bound
* Vendor `HMock` and `explainable-predicates`

## 0.6.0.0 (2023-05-28)

### Breaking

* Add support for `optparse-applicative` `0.18`

### Non-Breaking

* Bump `ansi-wl-pprint` dependency version upper bound
* Adjust dependency constraints to match tested versions

## 0.5.0.0 (2022-03-01)

### Breaking

* Rename Haskell package to `horizontal-rule`
* Refactor API to use `MonadTerminal` instead of `IO` directly

### Non-Breaking

* Add mock tests
* Bump `text` dependency version upper bound
* Bump `time` dependency version upper bound
* Bump `optparse-applicative` dependency version upper bound

## 0.4.0.0 (2021-06-25)

### Breaking

* Fix `--help` when using `optparse-applicative` `0.16`

### Non-Breaking

* Refactor Nix configuration

## 0.3.0.1 (2021-05-28)

### Non-Breaking

* Add library usage example

## 0.3.0.0 (2021-05-28)

### Breaking

* Add library
* Add `--width` and `--default` CLI options
* Add `--input` and `--timeout` CLI options

## 0.2.0.1 (2021-05-27)

### Non-Breaking

* Fix link in README
* Fix formatting in RPM description

## 0.2.0.0 (2021-05-26)

### Breaking

* Add support for `optparse-applicative` `0.16`

### Non-Breaking

* Add `.deb` and `.rpm` packaging
* Add Cabal support to `Makefile`
* Add Cabal tests to GitHub Actions
* Add [stan](https://hackage.haskell.org/package/stan) static analysis

## 0.1.0.3 (2020-11-21)

### Non-Breaking

* Use GitHub Actions instead of Travis CI

## 0.1.0.2 (2020-11-05)

### Non-Breaking

* Rename Git default branch to `main`
* Refactor `Makefile`, add `STACK_NIX_PATH` support
* Add `test-all` command to `Makefile`
* Add Nix configuration

## 0.1.0.1 (2019-12-29)

### Non-Breaking

* Fix operator for compatibility with GHC 8.2.2

## 0.1.0.0 (2019-12-29)

### Breaking

* Initial public release


hr.hs
------------------------------------------------------------------------------
-- |
-- Module      : Main
-- Description : hr: a horizontal rule for terminals
-- Copyright   : Copyright (c) 2019-2021 Travis Cardwell
-- License     : MIT
--
-- See the README for details.
------------------------------------------------------------------------------

{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE RecordWildCards #-}

module Main (main) where

-- https://hackage.haskell.org/package/ansi-wl-pprint
import Text.PrettyPrint.ANSI.Leijen (Doc)

-- https://hackage.haskell.org/package/base
import Control.Applicative (many)
import Data.List (intercalate)
import Data.Maybe (maybeToList)
import Data.Version (showVersion)

-- https://hackage.haskell.org/package/optparse-applicative
import qualified Options.Applicative as OA

-- https://hackage.haskell.org/package/terminal-size
import qualified System.Console.Terminal.Size as Terminal

-- https://hackage.haskell.org/package/time
import Data.Time.Format (defaultTimeLocale, formatTime)
import Data.Time.LocalTime (getZonedTime)

-- (hr:cabal)
import qualified Paths_hr as Project

-- (hr:executable)
import qualified LibOA

------------------------------------------------------------------------------

-- | Default time format
defaultFormat :: String
defaultFormat = "%Y-%m-%d %H:%M:%S"

-- | Default terminal width
defaultWidth :: Int
defaultWidth = 80

------------------------------------------------------------------------------

-- | Render a horizontal rule
hr :: Bool -> Int -> [String] -> String
hr ascii width = \case
    [] -> replicate width rule_fill
    notes ->
      let rule = rule_left ++ intercalate rule_mid notes ++ rule_right
          rule_len = length rule
      in  if rule_len < width
            then rule ++ replicate (width - rule_len) rule_fill
            else rule
  where
    rule_left = if ascii then "--|" else "━━┫"
    rule_mid = if ascii then "|-|" else "┣━┫"
    rule_right = if ascii then "|--" else "┣━━"
    rule_fill = if ascii then '-' else '━'

------------------------------------------------------------------------------

-- | Command-line options
data Options
  = Options
    { optAscii  :: !Bool
    , optTime   :: !Bool
    , optFormat :: !String
    , optNote   :: !(Maybe String)
    }
  deriving Show

options :: OA.Parser Options
options =
    Options
      <$> asciiOption
      <*> timeOption
      <*> formatOption
      <*> noteArguments

asciiOption :: OA.Parser Bool
asciiOption = OA.switch $ mconcat
    [ OA.long "ascii"
    , OA.short 'a'
    , OA.help "use ASCII lines"
    ]

timeOption :: OA.Parser Bool
timeOption = OA.switch $ mconcat
    [ OA.long "time"
    , OA.short 't'
    , OA.help "show time"
    ]

formatOption :: OA.Parser String
formatOption = OA.strOption $ mconcat
    [ OA.long "format"
    , OA.short 'f'
    , OA.metavar "FORMAT"
    , OA.value defaultFormat
    , OA.showDefaultWith id
    , OA.help "time format"
    ]

noteArguments :: OA.Parser (Maybe String)
noteArguments = fmap mjoin . many . OA.strArgument $ mconcat
    [ OA.metavar "NOTE ..."
    , OA.help "show a note"
    ]
  where
    mjoin :: [String] -> Maybe String
    mjoin [] = Nothing
    mjoin ss = Just $ unwords ss

------------------------------------------------------------------------------

main :: IO ()
main = do
    Options{..} <- OA.execParser pinfo
    mTime <- if optTime
      then Just . formatTime defaultTimeLocale optFormat <$> getZonedTime
      else pure Nothing
    width <- maybe defaultWidth Terminal.width <$> Terminal.size
    putStrLn . hr optAscii width $ maybeToList mTime ++ maybeToList optNote
  where
    pinfo :: OA.ParserInfo Options
    pinfo
      = OA.info (LibOA.helper <*> LibOA.versioner version <*> options)
      $ mconcat
          [ OA.fullDesc
          , OA.progDesc "horizontal rule for the terminal"
          , OA.failureCode 2
          , OA.footerDoc $ Just formatHelp
          ]

    version :: String
    version = "hr-haskell " ++ showVersion Project.version

    formatHelp :: Doc
    formatHelp = LibOA.section "FORMAT codes:" $ LibOA.table
      [ ("%Y", "four-digit year")
      , ("%y", "two-digit year")
      , ("%m", "two-digit month")
      , ("%d", "two-digit day")
      , ("%H", "two-digit hour using 24-hour clock")
      , ("%I", "two-digit hour using 12-hour clock")
      , ("%p", "locale equivalent of AM or PM")
      , ("%M", "two-digit minute")
      , ("%S", "two-digit second")
      , ("%f", "six-digit microsecond")
      , ("%z", "UTC offset")
      ]
